### PR TITLE
[geocoder] Add borders class to support region borders in reverse serverside geocoder

### PR DESCRIPTION
**Pull request by tatiana-yan**

### CODE REVIEW

**Comment by tatiana-yan** [comment]
JTTP

File: indexer/borders.cpp [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+#include "base/logging.hpp"
+#include "base/macros.hpp"
+
+#include "defines.hpp"
+
+using namespace std;
+
+namespace
+{
+template <class ReaderT>
```
Comment (mpimenov) [inline indexer/borders.cpp]: s/ReaderT/Reader/

File: indexer/borders.cpp [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+class BordersVector
+{
+  DISALLOW_COPY(BordersVector);
```
Comment (mpimenov) [inline indexer/borders.cpp]: This is usually the last line, not the first.

File: indexer/borders.hpp [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+#include <list>
+#include <map>
+#include <string>
+
+namespace indexer
+{
+class Borders
+{
+public:
+  bool IsPointInside(uint64_t id, m2::PointD point) const
```
Comment (mpimenov) [inline indexer/borders.hpp]: please move to .cpp

File: indexer/borders.hpp [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    return false;
+  }
+
+  bool Deserialize(std::string filename);
+
+  template <typename BordersVec>
+  void DeserializeFromVec(BordersVec const & vec)
+  {
+    vec.ForEach([this](uint64_t id, std::vector<m2::PointD> const & outer,
+                       std::list<std::vector<m2::PointD>> const & inners) {
```
Comment (mpimenov) [inline indexer/borders.hpp]: why std::list?

File: indexer/borders.hpp [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+      it->second.m_outer = m2::RegionD(outer);
+      for (auto const & inner : inners)
+        it->second.m_inners.push_back(m2::RegionD(inner));
+    });
+  }
+
+private:
+  struct Border
+  {
+    bool IsPointInside(m2::PointD point) const
```
Comment (mpimenov) [inline indexer/borders.hpp]: const & point

File: indexer/borders.hpp [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    {
```
Comment (mpimenov) [inline indexer/borders.hpp]: please move to .cpp

File: indexer/borders.cpp [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+private:
+  FilesContainerR m_cont;
+  BordersVector<ModelReaderPtr> m_vector;
+};
+}  // namespace
+
+namespace indexer
+{
+bool Borders::Deserialize(string filename)
```
Comment (mpimenov) [inline indexer/borders.cpp]: string const &

File: indexer/borders.cpp [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+public:
+  BordersVector(ReaderT const & reader) : m_recordReader(reader, 256 /* expectedRecordSize */) {}
+
+  template <class ToDo>
+  void ForEach(ToDo && toDo) const
+  {
+    m_recordReader.ForEachRecord([&](uint32_t pos, char const * data, uint32_t /*size*/) {
+      ArrayByteSource src(data);
+      serial::GeometryCodingParams cp = {};
```
Comment (mpimenov) [inline indexer/borders.cpp]: This should be read from the header, shouldn't it?

File: indexer/borders.cpp [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+      uint64_t id;
+      ReadPrimitiveFromSource(src, id);
+      size_t size;
+      ReadPrimitiveFromSource(src, size);
+      size_t outerSize;
+      ReadPrimitiveFromSource(src, outerSize);
+      vector<m2::PointD> outer(outerSize);
+      // todo:(@t.yan) consider delta coding techniques.
+      for (size_t i = 0; i < outer.size(); ++i)
```
Comment (mpimenov) [inline indexer/borders.cpp]: nit: if you don't use reserve+emplace_back, you can write range-based for here instead

File: indexer/borders.cpp (hunk 1 above)
Comment (tatiana-yan) [reply indexer/borders.cpp]: done

File: indexer/borders.cpp (hunk 2 above)
Comment (tatiana-yan) [reply indexer/borders.cpp]: done

File: indexer/borders.hpp (hunk 3 above)
Comment (tatiana-yan) [reply indexer/borders.hpp]: done

File: indexer/borders.hpp (hunk 5 above)
Comment (tatiana-yan) [reply indexer/borders.hpp]: done

File: indexer/borders.hpp (hunk 6 above)
Comment (tatiana-yan) [reply indexer/borders.hpp]: done

File: indexer/borders.cpp (hunk 8 above)
Comment (tatiana-yan) [reply indexer/borders.cpp]: пока что генерация в несколько разобранном состоянии и хедера там нет:) по мере работы над генерацией буду менять и этот код -- чтение параметров из хедера, дельта кодирование

File: indexer/borders.cpp (hunk 9 above)
Comment (tatiana-yan) [reply indexer/borders.cpp]: done

**Comment by tatiana-yan** [comment]
JTTP

File: indexer/borders.hpp (hunk 4 above)
Comment (tatiana-yan) [reply indexer/borders.hpp]: ok, vector

File: indexer/borders.cpp (hunk 7 above)
Comment (tatiana-yan) [reply indexer/borders.cpp]: done

File: indexer/borders.hpp [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+#include <cstdint>
+#include <map>
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace indexer
+{
+class Borders
```
Comment (mpimenov) [inline indexer/borders.hpp]: Please add a comment to this class: what it holds, where it is supposed to be used and what the inner structure is . The meaning of inner/outer is not immediately obvious.

Comment (tatiana-yan) [reply indexer/borders.hpp]: done

**Pull request merged**